--- ospfd/ospfd.h.orig	2005-11-15 11:32:31.000000000 +0100
+++ ospfd/ospfd.h	2005-11-15 11:33:58.000000000 +0100
@@ -371,9 +371,9 @@
   {
     char *name;
     struct access_list *list;
-  } export;
-#define EXPORT_NAME(A)  (A)->export.name
-#define EXPORT_LIST(A)  (A)->export.list
+  } export__item;
+#define EXPORT_NAME(A)  (A)->export__item.name
+#define EXPORT_LIST(A)  (A)->export__item.list
 
   /* Area acceptance list. */
   struct 
@@ -522,7 +522,7 @@
 extern int ospf_network_type_msg_max;
 extern struct zclient *zclient;
 extern struct thread_master *master;
-extern int ospf_zlog;
+// ospf_zlog not used anywhere
 
 /* Prototypes. */
 struct ospf *ospf_lookup ();
