--- lib/regex.c.orig	2004-11-24 17:14:49.000000000 +0000
+++ lib/regex.c	2006-10-21 13:51:44.000000000 +0000
@@ -27,6 +27,8 @@
 #undef	_GNU_SOURCE
 #define _GNU_SOURCE
 
+#include "globalvars.h"
+
 #ifdef HAVE_CONFIG_H
 # include <config.h>
 #endif
@@ -164,9 +166,6 @@
 #  define SWITCH_ENUM_CAST(x) (x)
 # endif
 
-/* How many characters in the character set.  */
-# define CHAR_SET_SIZE 256
-
 # ifdef SYNTAX_TABLE
 
 extern char *re_syntax_table;
@@ -1014,7 +1013,6 @@
    but why not be nice?  */
 
 static const char re_error_msgid[] =
-  {
 #define REG_NOERROR_IDX	0
     gettext_noop ("Success")	/* REG_NOERROR */
     "\0"
@@ -1025,7 +1023,7 @@
     gettext_noop ("Invalid regular expression") /* REG_BADPAT */
     "\0"
 #define REG_ECOLLATE_IDX (REG_BADPAT_IDX + sizeof "Invalid regular expression")
-    gettext_noop ("Invalid collation character"), /* REG_ECOLLATE */
+    gettext_noop ("Invalid collation character") /* REG_ECOLLATE */
     "\0"
 #define REG_ECTYPE_IDX	(REG_ECOLLATE_IDX + sizeof "Invalid collation character")
     gettext_noop ("Invalid character class name") /* REG_ECTYPE */
@@ -1064,8 +1062,7 @@
     gettext_noop ("Regular expression too big") /* REG_ESIZE */
     "\0"
 #define REG_ERPAREN_IDX	(REG_ESIZE_IDX + sizeof "Regular expression too big")
-    gettext_noop ("Unmatched ) or \\)"), /* REG_ERPAREN */
-  };
+    gettext_noop ("Unmatched ) or \\)"); /* REG_ERPAREN */
 
 static const size_t re_error_msgid_idx[] =
   {
@@ -1178,6 +1175,10 @@
 int re_max_failures = 2000;
 # endif
 
+// doxygen outputs wrong line number for this variable
+// moving it here helps, don't ask me why
+static char reg_unset_dummy;
+
 union fail_stack_elt
 {
   unsigned char *pointer;
@@ -1548,7 +1549,6 @@
   while (0)
 
 /* Registers are set to a sentinel when they haven't yet matched.  */
-static char reg_unset_dummy;
 #define REG_UNSET_VALUE (&reg_unset_dummy)
 #define REG_UNSET(e) ((e) == REG_UNSET_VALUE)
 
