--- ospfd/ospf_apiserver.c.orig	2005-11-13 14:53:07.000000000 +0100
+++ ospfd/ospf_apiserver.c	2005-11-13 14:53:58.000000000 +0100
@@ -301,7 +301,7 @@
 }
 
 void
-ospf_apiserver_event (enum event event, int fd,
+ospf_apiserver_event (enum ospfd_event event, int fd,
 		      struct ospf_apiserver *apiserv)
 {
   struct thread *apiserver_serv_thread;
@@ -417,7 +417,7 @@
   struct msg *msg;
   int fd;
   int rc = -1;
-  enum event event;
+  enum ospfd_event event;
 
   apiserv = THREAD_ARG (thread);
   fd = THREAD_FD (thread);
@@ -780,7 +780,7 @@
 {
   struct msg_fifo *fifo;
   struct msg *msg2;
-  enum event event;
+  enum ospfd_event event;
   int fd;
 
   switch (msg->hdr.msgtype)
